[turner.lee] 지하철 admin step4 리뷰 부탁드립니다.

sumniy opened the pull request:
안녕하세요 제임스. 
말씀해주신 부분 반영하고 step4의 리팩토링도 적용했습니다.

감사합니다.

sumniy committed:
Refactor: Station 리팩토링

Station에 Service 계층을 추가하여 Controller와 Dao 사이의
의존성을 분리함.
그리고 Dao는 순수하게 DB 접근만을 책임지게하고 그 결과에 대한 확인과
예외처리는 Service에서 담당하고 클라이언트에 보내는
예외 처리 응답은 Controller의 ExceptionHandler에게 위임.

sumniy committed:
Refactor: Line 리팩토링

LineService를 추가하여 Controller와 Dao 간의 의존성을 분리.
예외 처리는 Service에서 담당하고 throw한 예외는 Controller의
ExceptionHandler에서 응답하게 함.

sumniy committed:
Refactor: Section 리팩토링

SectionService에서 List<Section>에 적용되는
비즈니스 로직을 Sections 클래스가 담당하도록 변경.

sumniy committed:
Refactor: Line Controller 리팩토링

LineController가 가지고 있던 stationDao와의 의존관계를 없애고
stationService로 대체함.

sumniy committed:
Refactor: Station 예외 처리 개선

기존에는 stationDao가 중복된 이름의 station을 insert할 때 발생하는
exception을 처리해주는 방식으로 사용했는데, insert를 시도하는 것 자체의
비용이 크므로 select 쿼리로 존재여부를 판단하여 예외 처리를 하도록 변경함.

sumniy committed:
Refactor: Line 예외 처리 개선

station에서의 리팩토링과 비슷하게 insert, delete, update 등의
write가 필요한 쿼리전에 select 쿼리로 먼저 확인 후 예외 처리를 하도록 변경함.

sumniy committed:
Refactor: Section 예외 처리 개선

station에서의 리팩토링과 비슷하게 insert, delete, update 등의
write가 필요한 쿼리전에 select 쿼리로 먼저 확인 후 예외 처리를 하도록 변경함.
- sections 최소 크기 예외 처리 추가
- 삽입 시 비정상적인 상황에 대한 예외 처리 추가

sumniy committed:
Refactor: domain 객체와 request 객체의 의존 관계 수정

기존에 domain 객체에서 request를 의존하는 관계로 생성자를
구현했는데, 변화가 좀 더 잦을 수 있는 request 객체에서
domain을 의존하는 식, 즉 반대 방향으로 의존 관계를 갖게 하는게
더 적절하다고 판단되어 수정함.

findstar commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private void checkDuplicateName(String lineName) {
+        if (lineDao.findByName(lineName) == null) {
```
LineDao 에 동일한 이름이 존재하는지 확인하여 boolean 으로 반환하는 메서드를 추가하면 좀 더 간단해질것 같습니다.
그리고 == null 으로 체크하면 잘못된것 아닌가요?

sumniy replied:
넵 boolean을 반환하는 메서드로 수정해보겠습니다. 그런데 제가 LineDao의 findByName에서 name으로 조회된 결과가 없으면 null을 반환하게 만들어놓아서 == null로 체크했는데, 이 방법이 어떤 문제가 발생할 여지가 있을까요?

findstar commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public Line find(Long id) {
+        return Optional.ofNullable(lineDao.findById(id))
```
optional 을 사용하셨네요 💯 
그런데 이경우에는 Dao 에서 Optional<Line> 을 반환하면 어떨까요? Dao 에서 계속 null 관련 로직이 들어가는데 
이를 호출하는 곳에서도 null 을 신경써줘야 합니다.

sumniy replied:
위에서 언급해주신대로 LineDao에서 동일한 이름이 존재하는지, 동일한 id의 line이 존재하는지 체크하는 메서드를 만드니, 
dao에서 null 관련 로직을 없앨 수 있었습니다. 그래서 Optional도 사용하지 않게 되었습니다. 😅

findstar commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public Line find(Long id) {
+        return Optional.ofNullable(lineDao.findById(id))
```
optional 을 사용하셨네요 💯 
그런데 이경우에는 Dao 에서 Optional<Line> 을 반환하면 어떨까요? Dao 에서 계속 null 관련 로직이 들어가는데 
이를 호출하는 곳에서도 null 을 신경써줘야 합니다.

findstar commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,70 @@
+
+    public void delete(Long id) {
+        try {
+            checkExistLine(id);
```
delete 할 때 굳이 checkExist 하지 않아도 됩니다.

sumniy replied:
delete도 write 작업인데 그전에 미리 체크 해주는 것이 낫지 않나요? 아니면 delete 자체가 생성이나 수정에 비해서 많이 발생하지 않으니 큰 의미가 없는 것일까요? 🤔

findstar commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,70 @@
+
+    public Line find(Long id) {
+        return Optional.ofNullable(lineDao.findById(id))
+                .orElseThrow(() -> new NoSuchStationException(id));
```
orElseThrow 가 NoSuchStationException 을 반환하네요. NoSuchLineException 이 되어야 할것 같네요.
관례적으로 이런 경우
NotFoundLineException 과 같은 이름을 붙입니다 😄

sumniy replied:
감사합니다 ㅎㅎ

findstar commented on src/main/java/subway/section/SectionController.java:
```diff
@@ -1,45 +1,41 @@
 
 @RestController
 @RequestMapping("/lines/{lineId}")
 public class SectionController {
```
/lines/{lineId} 아래 sections 만 처리하므로 LineSectionController 로 이름을 바꿔보는건 어떤가요?

sumniy replied:
조금 애매하다 생각했던 부분인데 감사합니다!

findstar commented on src/main/java/subway/section/SectionController.java:
```diff
@@ -1,45 +1,41 @@
-        sectionService.addSectionToLine(Section.of(lineId, sectionRequest));
-        return ResponseEntity.ok().body(LineResponse.of(lineDao.findById(lineId)));
+        sectionService.addSectionToLine(sectionRequest.toDomainObject(lineId));
+        return ResponseEntity.ok().body(LineResponse.of(lineService.find(lineId)));
```
생성 이후 javascript 의 요구사항이 없다면 심플하게 ok().build() 처리해도 됩니다 😄

findstar commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -27,40 +24,22 @@
     @Transactional
     public void addSectionToLine(Section newSection) {
-        sections = sectionDao.getSectionsByLineId(newSection.getLineId());
+        Sections sections = new Sections(sectionDao.findSectionsByLineId(newSection.getLineId()));
```
일급 객체를 활용하도록 개선되었네요 💯

findstar commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -27,40 +24,22 @@
-                .anyMatch(section -> section.isSameSection(newSection))) {
-            throw new IllegalArgumentException("같은 구역이 이미 등록되어 있습니다.");
-        }
+        addInMiddle(sections, newSection);
```
복잡한 로직을 분리해서 처리할 수 있게되면서 addSectionInLine 메서드가 간결해졌습니다 💯

findstar reviewed (changes requested):
안녕하세요 제임스입니다.
지난번 피드백 잘 반영해서 리팩토링 진행해주셨네요 👍
몇가지 추가로 코멘트 남겼습니다. 확인 부탁드립니다. 
계속 화이팅입니다! 궁금한점은 언제든지 DM 주세요 😄

sumniy committed:
Refactor: 피드백 반영

- Dao에 boolean을 반환하는 존재여부 관련 메서드를 추가하여, find 메서드에 null 관련 로직을 없애고
호출하는 메서드에서도 간단하게 처리가 가능해짐.
- 네이밍 변경

sumniy commented:
안녕하세요 제임스! 터너입니다. 
남겨주신 피드백 반영하였고, 궁금한 점들 댓글로 남겨놨습니다. 

추가로 궁금한 점이 있습니다. 
저는 컨트롤러에서 request를 deserialize하고 response를 serialize하도록 해서 Service는 도메인 객체와 의존성을 갖도록 했습니다. 
이렇게 하면 컨트롤러가 깔끔하지는 않지만 Service 객체에 dto 관련 책임이 없어져서 유지 보수나 코드의 재사용성이 더 좋을 거라 생각했는데요. 
제임스는 어떻게 생각하시는지 궁금합니다. 

감사합니다.

findstar reviewed (approved):
안녕하세요 제임스입니다. 
코드 수정 잘 해주셨네요 💯 
질문 주신 내용에 대해서 제가 생각하는 부분을 답변드리면 
저는 컨트롤러에서 request를 바로 서비스 객체로 던지지 않고 
도메인 객체나 DTO 를 별도로 생성해서 서비스 객체에서 처리하도록 하고 있습니다.
Production 에서는 Request 객체가 모두 동일한 형태를 가지기 어렵기 때문에
Request 객체를 서비스에서 바로 전달하는 구조는 지양하고 있습니다. 
더 궁금한 내용이 있다면 언제든 DM 주세요 😄 
미션 수행하시느라 수고 많으셨습니다.

sumniy commented:
감사합니다!
